diff --git a/include/spdlog/details/periodic_worker.h b/include/spdlog/details/periodic_worker.h
index d05245cb..3252e53a 100644
--- a/include/spdlog/details/periodic_worker.h
+++ b/include/spdlog/details/periodic_worker.h
@@ -21,14 +21,16 @@ namespace details {
 class SPDLOG_API periodic_worker {
 public:
     template <typename Rep, typename Period>
-    periodic_worker(const std::function<void()> &callback_fun,
+    periodic_worker(const std::function<void()> &init_thread_fun,
+                    const std::function<void()> &callback_fun,
                     std::chrono::duration<Rep, Period> interval) {
         active_ = (interval > std::chrono::duration<Rep, Period>::zero());
         if (!active_) {
             return;
         }
 
-        worker_thread_ = std::thread([this, callback_fun, interval]() {
+        worker_thread_ = std::thread([this, init_thread_fun, callback_fun, interval]() {
+            init_thread_fun();
             for (;;) {
                 std::unique_lock<std::mutex> lock(this->mutex_);
                 if (this->cv_.wait_for(lock, interval, [this] { return !this->active_; })) {
