diff --git a/CMakeLists.txt b/CMakeLists.txt
index ccba973..8b2cbf2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -70,6 +70,11 @@ elseif (CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
     message(STATUS "CMAKE_CXX_COMPILER_VERSION: ${CMAKE_CXX_COMPILER_VERSION}")
     message(FATAL_ERROR "\nTaskflow requires icpc at least v19.0.1")
   endif()
+elseif (CMAKE_CXX_COMPILER_ID STREQUAL "IntelLLVM")
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "2021.1.0")
+    message(STATUS "CMAKE_CXX_COMPILER_VERSION: ${CMAKE_CXX_COMPILER_VERSION}")
+    message(FATAL_ERROR "\nTaskflow requires icpx at least v2021.1.0")
+  endif()
 else()
   message(FATAL_ERROR "\n\
 Taskflow currently supports the following compilers:\n\
@@ -78,6 +83,7 @@ Taskflow currently supports the following compilers:\n\
   - MSVC++ v19.14 or above\n\
   - AppleClang v8 or above\n\
   - Intel v19.0.1 or above\n\
+  - IntelLLVM v2021.1.0 or above\n\
 ")
 endif()
 
@@ -372,4 +378,3 @@ install(
         ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake
   DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
 )
-
